consensus/consensus-types/src/block.rs

```diff
@@ -300,6 +302,47 @@ impl Block {
             )
             .collect()
     }
+
+    fn new_block_metadata(&self, validators: &Vec<AccountAddress>) -> BlockMetadata {
+        let participants: HashSet<_> = self
+            .quorum_cert()
+            .ledger_info()
+            .signatures()
+            .keys()
+            .collect();
+
+        BlockMetadata::new(
+            self.id(),
+            self.epoch(),
+            self.round(),
+            self.voters_to_bitmap(validators, participants),
+            self.proposer_to_index(validators, self.author()),
+            self.timestamp_usecs(),
+        )
+    }
+
+    // TODO: as util function that can be tested with the reverse operation.
+    // TODO: also see functions in leader_reputation
+    // A bitmap of previous blocks, starting at 1
+    fn voters_to_bitmap(&self, validators: &Vec<AccountAddress>, voters: HashSet<&AccountAddress>) -> Vec<bool> {
+        std::iter::once(false)
+            .chain(
+                validators
+                    .iter()
+                    .map(|address| voters.contains(address))
+            )
+            .collect::<Vec<_>>()
+    }
+
+    // TODO: as util function that can be tested with the reverse operation
+    // TODO: also see functions in leader_reputation
+    // For nil block, we use 0
+    fn proposer_to_index(&self, validators: &Vec<AccountAddress>, proposer: Option<AccountAddress>) -> u64 {
+        (match proposer {
+            Some(address) => validators.iter().find_position(|a| **a == address).unwrap_or((0, &AccountAddress::ZERO)).0,
```

Comment:
position instead of find_position? also we need to differentiate not found from index 0, is the index supposed to +1?